xarray/core/common.py

```diff
@@ -705,54 +705,69 @@ def _get_fill_value(dtype):
-            `func` is calculated over all dimension for each group item.
-    keep_attrs : bool, optional
-        If True, the datasets's attributes (`attrs`) will be copied from
-        the original object to the new one.  If False (default), the new
-        object will be returned without attributes.
-    fill_value : scalar, optional
-        Value to fill DataArray(s) with before returning.
+    other : DataArray, Dataset, or Variable
+        The reference object in input
+    fill_value : scalar
+        Value to fill the new object with before returning it.
+    dtype : dtype, optional
+        dtype of the new array. If omitted, it defaults to other.dtype.
 
     Returns
     -------
-    out : same as xarray_obj
-        New object with the same shape and type as a given object.
+    out : same as object
+        New object with the same shape and type as other, with the data
+        filled with fill_value. Coords will be copied from other.
+        If other is based on dask, the new one will be as well, and will be
+        split in the same chunks.
     """
     from .dataarray import DataArray
     from .dataset import Dataset
+    from .variable import Variable
+
+    if isinstance(other, Dataset):
+        data_vars = OrderedDict(
+            (k, _full_like_variable(v, fill_value, dtype))
+            for k, v in other.data_vars.items())
+        return Dataset(data_vars, coords=other.coords, attrs=other.attrs)
+    elif isinstance(other, DataArray):
+        return DataArray(
+            _full_like_variable(other.variable, fill_value, dtype),
+            dims=other.dims, coords=other.coords, attrs=other.attrs, name=other.name)
+    elif isinstance(other, Variable):
+        return _full_like_variable(other, fill_value, dtype)
+    else:
+        raise TypeError("Expected DataArray, Dataset, or Variable")
 
-    if isinstance(xarray_obj, Dataset):
-        attrs = xarray_obj.attrs if keep_attrs else {}
 
-        return Dataset(dict((k, _full_like_dataarray(v, keep_attrs=keep_attrs,
-                                                     fill_value=fill_value))
-                            for k, v in iteritems(xarray_obj.data_vars)),
-                       name=xarray_obj.name, attrs=attrs)
-    elif isinstance(xarray_obj, DataArray):
-        return _full_like_dataarray(xarray_obj, keep_attrs=keep_attrs,
-                                    fill_value=fill_value)
+def _full_like_variable(other, fill_value, dtype=None):
+    """Inner function of full_like, where other must be a variable
+    """
+    from .variable import Variable
+
+    if isinstance(other.data, dask_array_type):
+        import dask.array
+        if dtype is None:
+            dtype = other.dtype
+        data = dask.array.full(other.shape, fill_value, dtype=dtype,
+                               chunks=other.data.chunks)
+    else:
+        data = np.full_like(other, fill_value, dtype=dtype)
```

Comment:
This works, but it might be marginally safer to pass in `other.data` rather than relying on NumPy properly handling xarray's array-like `Variable` type.